Compilation failure with ffmpeg 5.0

## Environment
* OS Version: macOS 10.15
* Source or binary build? source build all major versions 1-5

## Description

ffmpeg 5.0 has just been merged in https://github.com/Homebrew/homebrew-core/pull/94423

* Expected behavior: ignition-common compiles with ffmpeg
* Actual behavior: it fails to compile

## Steps to reproduce

1. `brew install ignition-common5`
2. observer build failure

## Output

~~~
cd /tmp/ignition-common5-20220203-58814-14kxgm5/events/src && /usr/local/Cellar/cmake/3.22.2/bin/cmake -E cmake_link_script CMakeFiles/ignition-common5-events.dir/link.txt --verbose=1
/tmp/ignition-common5-20220203-58814-14kxgm5/av/src/VideoEncoder.cc:132:10: error: cannot initialize return object of type 'AVCodec *' with an rvalue of type 'const AVCodec *'
  return avcodec_find_encoder(_codecId);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++  -stdlib=libc++ -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -dynamiclib -Wl,-headerpad_max_install_names  -Wl,-undefined -Wl,dynamic_lookup -compatibility_version 5.0.0 -current_version 5.0.0 -o "../../lib/libignition-common5-events.5.0.0~pre1.dylib" -install_name @rpath/libignition-common5-events.5.dylib CMakeFiles/ignition-common5-events.dir/Event.cc.o CMakeFiles/ignition-common5-events.dir/KeyEvent.cc.o CMakeFiles/ignition-common5-events.dir/MouseEvent.cc.o  -Wl,-rpath,/tmp/ignition-common5-20220203-58814-14kxgm5/lib "../../lib/libignition-common5.5.0.0~pre1.dylib" "/usr/local/lib/libignition-math7.7.0.0~pre1.dylib" /usr/local/lib/libignition-utils1.1.2.0.dylib -lpthread 
/tmp/ignition-common5-20220203-58814-14kxgm5/av/src/VideoEncoder.cc:370:21: error: cannot initialize a variable of type 'AVOutputFormat *' with an rvalue of type 'const AVOutputFormat *'
    AVOutputFormat *outputFormat = av_guess_format(nullptr,
                    ^              ~~~~~~~~~~~~~~~~~~~~~~~~
[ 66%] Building CXX object graphics/src/CMakeFiles/ignition-common5-graphics.dir/Image.cc.o
cd /tmp/ignition-common5-20220203-58814-14kxgm5/graphics/src && /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++ -DMAC_OS_X_VERSION=1080 -DTINYXML2_MAJOR_VERSION_GE_6 -Dignition_common5_graphics_EXPORTS -I/tmp/ignition-common5-20220203-58814-14kxgm5/include -I/tmp/ignition-common5-20220203-58814-14kxgm5/graphics/include -I/tmp/ignition-common5-20220203-58814-14kxgm5/core/include -isystem /usr/local/include/ignition/utils1 -isystem /usr/local/include/ignition/math7 -isystem /usr/local/Cellar/pcre/8.45/include -isystem /usr/local/Cellar/glib/2.70.3/include -isystem /usr/local/Cellar/glib/2.70.3/include/glib-2.0 -isystem /usr/local/Cellar/glib/2.70.3/lib/glib-2.0/include -isystem /usr/local/opt/gettext/include -isystem /usr/local/Cellar/gts/0.7.6_2/include -isystem /usr/local/Cellar/tinyxml2/9.0.0/include -stdlib=libc++ -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -fPIC -fPIC -I/usr/local/Cellar/pcre/8.45/include -I/usr/local/Cellar/glib/2.70.3/include -I/usr/local/Cellar/glib/2.70.3/include/glib-2.0 -I/usr/local/Cellar/glib/2.70.3/lib/glib-2.0/include -I/usr/local/opt/gettext/include -I/usr/local/Cellar/gts/0.7.6_2/include -I/usr/local/Cellar/tinyxml2/9.0.0/include -std=c++17 -MD -MT graphics/src/CMakeFiles/ignition-common5-graphics.dir/Image.cc.o -MF CMakeFiles/ignition-common5-graphics.dir/Image.cc.o.d -o CMakeFiles/ignition-common5-graphics.dir/Image.cc.o -c /tmp/ignition-common5-20220203-58814-14kxgm5/graphics/src/Image.cc
cd /tmp/ignition-common5-20220203-58814-14kxgm5/events/src && /usr/local/Cellar/cmake/3.22.2/bin/cmake -E cmake_symlink_library "../../lib/libignition-common5-events.5.0.0~pre1.dylib" ../../lib/libignition-common5-events.5.dylib ../../lib/libignition-common5-events.dylib
2 errors generated.
[ 66%] Built target ignition-common5-events
make[2]: *** [av/src/CMakeFiles/ignition-common5-av.dir/VideoEncoder.cc.o] Error 1
~~~

--- a/av/src/AudioDecoder.cc
+++ b/av/src/AudioDecoder.cc
@@ -35,7 +35,7 @@ class ignition::common::AudioDecoderPrivate
   public: AVCodecContext *codecCtx;
 
   /// \brief libavcodec audio codec.
-  public: AVCodec *codec;
+  public: const AVCodec *codec;
 
   /// \brief Index of the audio stream.
   public: int audioStream;
@@ -77,8 +77,12 @@ void AudioDecoder::Cleanup()
 /////////////////////////////////////////////////
 bool AudioDecoder::Decode(uint8_t **_outBuffer, unsigned int *_outBufferSize)
 {
+#if LIBAVFORMAT_VERSION_MAJOR < 59
   AVPacket *packet, packet1;
   int bytesDecoded = 0;
+#else
+  AVPacket *packet;
+#endif
   unsigned int maxBufferSize = 0;
   AVFrame *decodedFrame = nullptr;
 
@@ -121,6 +125,53 @@ bool AudioDecoder::Decode(uint8_t **_outBuffer, unsigned int *_outBufferSize)
   {
     if (packet->stream_index == this->data->audioStream)
     {
+#if LIBAVFORMAT_VERSION_MAJOR >= 59
+      // Inspired from
+      // https://github.com/FFmpeg/FFmpeg/blob/n5.0/doc/examples/decode_audio.c#L71
+
+      // send the packet with the compressed data to the decoder
+      int ret = avcodec_send_packet(this->data->codecCtx, packet);
+      if (ret < 0)
+      {
+        ignerr << "Error submitting the packet to the decoder" << std::endl;
+        return false;
+      }
+
+      // read all the output frames
+      // (in general there may be any number of them)
+      while (ret >= 0)
+      {
+        ret = avcodec_receive_frame(this->data->codecCtx, decodedFrame);
+        if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF)
+        {
+          break;
+        }
+        else if (ret < 0)
+        {
+            ignerr << "Error during decoding" << std::endl;
+            return false;
+        }
+
+        // Total size of the data. Some padding can be added to
+        // decodedFrame->data[0], which is why we can't use
+        // decodedFrame->linesize[0].
+        int size = decodedFrame->nb_samples *
+          av_get_bytes_per_sample(this->data->codecCtx->sample_fmt) *
+          this->data->codecCtx->channels;
+
+        // Resize the audio buffer as necessary
+        if (*_outBufferSize + size > maxBufferSize)
+        {
+          maxBufferSize += size * 5;
+          *_outBuffer = reinterpret_cast<uint8_t*>(realloc(*_outBuffer,
+                maxBufferSize * sizeof(*_outBuffer[0])));
+        }
+
+        memcpy(*_outBuffer + *_outBufferSize, decodedFrame->data[0],
+            size);
+        *_outBufferSize += size;
+    }
+#else
       int gotFrame = 0;
 
       packet1 = *packet;
@@ -163,6 +214,7 @@ bool AudioDecoder::Decode(uint8_t **_outBuffer, unsigned int *_outBufferSize)
         packet1.data += bytesDecoded;
         packet1.size -= bytesDecoded;
       }
+#endif
     }
     av_packet_unref(packet);
   }
@@ -224,8 +276,13 @@ bool AudioDecoder::SetFile(const std::string &_filename)
 # pragma GCC diagnostic push
 # pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #endif
+#if LIBAVFORMAT_VERSION_MAJOR >= 59
+    if (this->data->formatCtx->streams[i]->codecpar->codec_type == // NOLINT(*)
+        AVMEDIA_TYPE_AUDIO)
+#else
     if (this->data->formatCtx->streams[i]->codec->codec_type == // NOLINT(*)
         AVMEDIA_TYPE_AUDIO)
+#endif
 #ifndef _WIN32
 # pragma GCC diagnostic pop
 #endif
@@ -249,14 +306,35 @@ bool AudioDecoder::SetFile(const std::string &_filename)
 # pragma GCC diagnostic push
 # pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #endif
+#if LIBAVFORMAT_VERSION_MAJOR < 59
   this->data->codecCtx = this->data->formatCtx->streams[
     this->data->audioStream]->codec;
+#endif
 #ifndef _WIN32
 # pragma GCC diagnostic pop
 #endif
 
   // Find a decoder
+#if LIBAVFORMAT_VERSION_MAJOR >= 59
+  this->data->codec = avcodec_find_decoder(this->data->formatCtx->streams[
+    this->data->audioStream]->codecpar->codec_id);
+  if (!this->data->codec)
+  {
+    ignerr << "Failed to find the codec" << std::endl;
+    return false;
+  }
+  this->data->codecCtx = avcodec_alloc_context3(this->data->codec);
+  if (!this->data->codecCtx)
+  {
+    ignerr << "Failed to allocate the codec context" << std::endl;
+    return false;
+  }
+  // Copy all relevant parameters from codepar to codecCtx
+  avcodec_parameters_to_context(this->data->codecCtx,
+    this->data->formatCtx->streams[this->data->audioStream]->codecpar);
+#else
   this->data->codec = avcodec_find_decoder(this->data->codecCtx->codec_id);
+#endif
 
   if (this->data->codec == nullptr)
   {

--- a/av/src/AudioDecoder_TEST.cc
+++ b/av/src/AudioDecoder_TEST.cc
@@ -128,8 +128,10 @@ TEST(AudioDecoder, IGN_UTILS_TEST_DISABLED_ON_WIN32(CheerFile))
     audio.Decode(&dataBuffer, &dataBufferSize);
     // In Ubuntu trusty the buffer size double for ogg decoding.
     // This check is suitable for both older and newer versions of Ubuntu.
+    // With ffmpeg 5.0 the value changed again (third value)
     EXPECT_TRUE(dataBufferSize == 4989184u ||
-                dataBufferSize == 4989184u * 2u);
+                dataBufferSize == 4989184u * 2u ||
+                dataBufferSize == 9975224u);
   }
 
   // MP3

--- a/av/src/Video.cc
+++ b/av/src/Video.cc
@@ -91,7 +91,7 @@ void Video::Cleanup()
 /////////////////////////////////////////////////
 bool Video::Load(const std::string &_filename)
 {
-  AVCodec *codec = nullptr;
+  const AVCodec * codec = nullptr;
   this->dataPtr->videoStream = -1;
 
   if (this->dataPtr->formatCtx || this->dataPtr->avFrame ||

--- a/av/src/VideoEncoder.cc
+++ b/av/src/VideoEncoder.cc
@@ -106,7 +106,7 @@ class IGNITION_COMMON_AV_HIDDEN ignition::common::VideoEncoderPrivate
   /// Find a suitable encoder for the given codec ID.
   /// \param[in] _codecId ID of the codec we seek the encoder for.
   /// \return The matched encoder (or nullptr on failure).
-  public: AVCodec* FindEncoder(AVCodecID _codecId);
+  public: const AVCodec* FindEncoder(AVCodecID _codecId);
 
   /// \brief Get a pointer to the frame that contains the encoder input. This
   /// mainly serves for uploading the frame to GPU buffer if HW acceleration is
@@ -123,7 +123,7 @@ class IGNITION_COMMON_AV_HIDDEN ignition::common::VideoEncoderPrivate
 };
 
 /////////////////////////////////////////////////
-AVCodec* VideoEncoderPrivate::FindEncoder(AVCodecID _codecId)
+const AVCodec* VideoEncoderPrivate::FindEncoder(AVCodecID _codecId)
 {
 #ifdef IGN_COMMON_BUILD_HW_VIDEO
   if (this->hwEncoder)
@@ -367,7 +367,7 @@ bool VideoEncoder::Start(
   }
   else
   {
-    AVOutputFormat *outputFormat = av_guess_format(nullptr,
+    const AVOutputFormat *outputFormat = av_guess_format(nullptr,
                                    this->dataPtr->filename.c_str(), nullptr);
 
     if (!outputFormat)

